Implementing proof of reserves

ulrichard opened the pull request:
### Description

Adding a module that allows the construction of "proof of reserves" PSBTs and the verification thereof.

### Notes to the reviewers

There will be an accompanying PR in bdk-cli. I will paste the URL here, once it's created.
https://github.com/ulrichard/bdk-cli/tree/reserves
But it builds upon this one:
https://github.com/bitcoindevkit/bdk-cli/pull/28

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [ ] I've added docs for the new feature
* [x] I've updated `CHANGELOG.md`

ulrichard committed:
Implementing proof of reserves

notmandatory commented on Cargo.toml:
```diff
@@ -67,6 +71,8 @@ env_logger = "0.7"
 clap = "2.33"
 serial_test = "0.4"
+rstest = "^0.7"
+tiny-bip39 = "^0.7"
```
The clippy error seems to be caused by this line, error goes away when I remove it. Could be because version `^0.8` is defined above (line 33) ?

ulrichard replied:
yes, you're right. Thanks for finding that.

notmandatory commented:
@nicolasburtey heard you on SLP and that you are looking into proof of reserve for your project. I would be interested to have your feedback on if you think this PR is in the right direction.

afilini commented on src/wallet/reserves.rs:
```diff
@@ -0,0 +1,603 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
```
I think this is our older header, see one of our other files for the new one

afilini commented on src/wallet/reserves.rs:
```diff
@@ -0,0 +1,603 @@
+    Network,
+};
+use bitcoin_hashes::{hash160, sha256d, Hash};
+use bitcoinconsensus;
```
We normally use the "module" import granularity, see https://rust-lang.github.io/rustfmt/?version=v1.4.36&search=impo#imports_granularity

ulrichard replied:
Would it make sense to add a rustfmt.toml or .rustfmt.toml file to the project?

afilini replied:
Yeah, that makes sense. I was assuming `cargo fmt` would enforce our version by default, but that's not the case

ulrichard replied:
Upon closer inspection, it might be too early. It's still an unstable feature in rustfmt.

afilini commented on src/wallet/reserves.rs:
```diff
@@ -0,0 +1,603 @@
+    /// implementation
+    fn do_create_proof(&self, message: &str) -> Result<PSBT, Error>;
+    /// implementation
+    fn do_verify_proof(&self, psbt: &PSBT, message: &str) -> Result<u64, Error>;
```
I don't understand why you would need those two other methods. Can't people just implement `create_proof` and `verify_proof` directly?

ulrichard replied:
That comes from 20 years of C++, where this was second nature. I'm still getting used to rust.
https://stackoverflow.com/questions/6481260/non-virtual-interface-design-pattern-in-c-c

afilini replied:
Conceptually it makes a lot of sense, but unfortunately in Rust you can prevent anybody from implementing a different version of your methods, even if you provide an implementation yourself.

This is generally used to provide a generic default implementation that people can then override with a different one if they want to

afilini commented on src/wallet/reserves.rs:
```diff
@@ -0,0 +1,603 @@
+        let mut builder = self.build_tx();
+        builder
+            .drain_wallet()
+            .add_foreign_utxo(challenge_txin.previous_output, challenge_psbt_inp, 42)?
```
If you need this to stay in a fixed position you probably need to set the ordering to `Untouched` explicitly, otherwise the inputs might get shuffled around: https://docs.rs/bdk/0.7.0/bdk/wallet/tx_builder/struct.TxBuilder.html#method.ordering

afilini commented on src/wallet/reserves.rs:
```diff
@@ -0,0 +1,603 @@
+            "Witness and legacy input found at input #{}",
+            i
+        )));
+    }
```
This is not really an error, the spec allows having both. Actually most signer nowadays require both for segwit inputs

afilini commented on src/wallet/reserves.rs:
```diff
@@ -0,0 +1,603 @@
+            "Unsupported sighash type at input #{}",
+            i
+        )));
+    }
```
Here you should check the sighash that actually ended up in the signature, not just the psbt metadata. Otherwise an attacker can make a signature with sighash X and then just say "no I did it with Y" in the metadata, which is not right.

ulrichard replied:
I didn't find this flag in the transaction, even after reading a couple of pages Ion the topic:
* https://raghavsood.com/blog/2018/06/10/bitcoin-signature-types-sighash
* https://docs.rs/bitcoin/0.26.0/bitcoin/blockdata/script/struct.Script.html
Can you guide me to the right direction a bit here?

ulrichard replied:
So, I wanted to verify SIGHASH type with something like:

```rust
if let Some((i, _inp)) = tx.input.iter().enumerate().skip(1).find(|(_i, inp)| {
        if let Some(sht) = inp.script_sig.as_bytes().last() {
            *sht != 0 && SigHashType::from_u32(*sht as u32) != SigHashType::All
        } else {
            true
        }
    }) {
        return Err(Error::Proof(ProofError::UnsupportedSighashType(i)));
    }
```

But the script_sig is only populated in the multisig test cases. In the single-sig cases, 
`psbt.clone().extract_tx().input[1].script_sig.as_bytes().len()` 
is always 0

What could be the reason for that, and how should I cope with that?

ulrichard replied:
Can I also interpret the last byte of the `witness: Vec<Vec<u8>>` as SigHashType?

ulrichard replied:
With the following function, most unit tests pass. But the code doesn't look correct to me.

```rust
...
let tx = psbt.clone().extract_tx();
// Verify the SIGHASH
if let Some((i, _inp)) = tx
        .input
        .iter()
        .enumerate()
        .skip(1)
        .find(|(_i, inp)| !verify_sighash_type_all(inp))
{
     return Err(Error::Proof(ProofError::UnsupportedSighashType(i)));
}
...
```

```rust
fn verify_sighash_type_all(inp: &TxIn) -> bool {
    if inp.witness.is_empty() {
        if let Some(sht) = inp.script_sig.as_bytes().last() {
            if SigHashType::from_u32(*sht as u32) == SigHashType::All {
                return true;
            } else {
                return false;
            }
        } else {
            return false;
        }
    } else {
        for wit in &inp.witness {
            if wit.last().is_none() {
                continue;
            }
            let sht = SigHashType::from_u32(*wit.last().unwrap() as u32);
            if ![
                SigHashType::All,
                SigHashType::Single,
                SigHashType::AllPlusAnyoneCanPay,
                SigHashType::SinglePlusAnyoneCanPay,
            ]
            .contains(&sht)
            {
                return false;
            }
        }
        return true;
    }
}
```
But still the legacy multisig case fails. It has an `sht` value of 0xAE which doesn't really translate to a valid SigHashType.

ulrichard replied:
I added some asserts temporarily to the unit tests to visualize the test data. 
They show a couple of things that I didn't expect from the information I found on the internet.
For the moment I adapted the verify_sighash_type_all() to let them pass, and added some comments.
I am curious to get some explanations on these values.

afilini commented on src/wallet/reserves.rs:
```diff
@@ -0,0 +1,603 @@
+                let outp = &nwtx.output[tx_in.previous_output.vout as usize];
+                (i, &outp.script_pubkey, outp.value)
+            } else {
+                panic!("must be either witness or legacy");
```
Ideally you shouldn't panic here, you can just return an error

afilini commented on src/wallet/reserves.rs:
```diff
@@ -0,0 +1,603 @@
+                (i, &utxo.script_pubkey, utxo.value)
+            } else if let Some(nwtx) = &psbt_in.non_witness_utxo {
+                let outp = &nwtx.output[tx_in.previous_output.vout as usize];
+                (i, &outp.script_pubkey, outp.value)
```
I think again here you are just trusting the value from the psbt metadata. This is dangerous: https://github.com/bitcoindevkit/bdk/issues/334#issuecomment-845059471

ulrichard replied:
Do I understand your comment correctly, that I should fetch the previous transaction as referenced in OutPoint to determine the value?

ulrichard replied:
Ok, now i'm passing in TxOut alongside OutPoint and looking up from there. That way I don't have to pass in an electrum client instance, and the blockchain lookup happens outside of the verification code.

afilini commented on src/wallet/reserves.rs:
```diff
@@ -0,0 +1,603 @@
+                nwtx.output[tx_in.previous_output.vout as usize].value
+            } else {
+                0
+            }
```
Same here, you are just trusting the `value` from the psbt input

afilini commented on src/wallet/reserves.rs:
```diff
@@ -0,0 +1,603 @@
+        previous_output: OutPoint::new(Txid::from_hash(message), 0),
+        sequence: 0xFFFFFFFF,
+        script_sig: Builder::new().into_script(),
+        witness: Vec::new(),
```
You can probably replace those two lines with:

```suggestion
        ..Default::default(),
```

afilini commented on src/error.rs:
```diff
@@ -64,6 +64,10 @@ pub enum Error {
+    /// The proof of reserves is invalid. The reason is given in the string parameter.
+    ProofOfReservesInvalid(String),
+    /// Cannot verify the proof.
+    CannotVerifyProof,
```
I would define a specific error enum for this, like we did with `SignerError` for instance. It's just to stop this `Error` from growing too much and providing some kind of a logical hierarchy for the errors

afilini replied:
Also with an isolated enum I would try to define a variant for every error instead of just having one that contains a `String`

afilini commented on Cargo.toml:
```diff
@@ -67,6 +71,7 @@ env_logger = "0.7"
 base64 = "^0.11"
 clap = "2.33"
 serial_test = "0.4"
+rstest = "^0.7"
```
I'm not a big fan of adding extra dependencies, even if they are only used during tests since they increase compilation times. Is there a different way to achieve the same result without adding this extra dependency?

ulrichard replied:
I'm sure it would be possible with some macro magic. But I don't know rust macros good enough yet.

afilini replied:
Ok, we can leave this open for the meantime and once everything else is done I'll help you out with the macro

afilini commented on Cargo.toml:
```diff
@@ -31,6 +31,9 @@ cc = { version = ">=1.0.64", optional = true }
 lazy_static = { version = "1.4", optional = true }
 tiny-bip39 = { version = "^0.8", optional = true }
+bitcoinconsensus = { version = "0.19.0-3", optional = true }
+bitcoin_hashes = { version = "^0.9", optional = true }
```
You don't have to add this as an extra deps, `bitcoin` re-exports it as `bitcoin::hashes` iirc

ulrichard committed:
Adding a separate error enum for proof of reserves

ulrichard committed:
also feeding the TxOut to verify_proof

ulrichard commented on src/wallet/reserves.rs:
```diff
@@ -0,0 +1,561 @@
+    network: Network,
+) -> Result<u64, Error> {
+    let tx = psbt.clone().extract_tx();
+
```
I'm not sure if it would make sense to copy the outpoints into a sorted container for faster lookups here.

nicolasburtey commented:
@notmandatory super cool I'll have a look. thanks.

notmandatory commented:
@ulrichard I have a meta question also about this PR. It seems that you don't need to access any wallet internals to create or validate a proof of reserve. So how would you feel about making this it's own repo here in the `bitcoindevkit` project? I think this would make it easier to iterate on the PoR concepts and release on your own schedule. And you can still have `pub extern crate bdk` so anyone using your crate will have full access to bdk wallet features.

ulrichard commented:
> @ulrichard I have a meta question also about this PR. It seems that you don't need to access any wallet internals to create or validate a proof of reserve. So how would you feel about making this it's own repo here in the `bitcoindevkit` project? I think this would make it easier to iterate on the PoR concepts and release on your own schedule. And you can still have `pub extern crate bdk` so anyone using your crate will have full access to bdk wallet features.

I'm still fairly new to rust, and thus still learning how to structure projects. If you think that's better, it's fine with me. 

1. Do you have naming conventions for stuff like that? How about "bdk-reserves"? 
2. Can I just create the project, and you add it to the `bitcoindevkit` project, or is it better if you create it first?

ulrichard committed:
First step for validating the SigHashType on the tx instead of psbt.
Made the verify_sighash_type_all() function work with the test data. But it does not all make sense to me.

notmandatory commented:
> I'm still fairly new to rust, and thus still learning how to structure projects. If you think that's better, it's fine with me.

Thanks, we're still figuring it out too, but I think in this case since proof of reserve is still an experimental topic it makes sense for it to live in it's own repo. 

>     1. Do you have naming conventions for stuff like that? How about "bdk-reserves"?

We don't have an official convention, the most important thing is that the name isn't taken in https://crates.io yet. The name `bdk-reserves` is a available. 

>     2. Can I just create the project, and you add it to the `bitcoindevkit` project, or is it better if you create it first?

It's probably easier if you create the repo in your account and then when you have it ready to re-review you can transfer it to the `bitcoindevkit` project. I don't think your code will need to change much, you should still be able to implementing the `ProofOfReserves` trait on `Wallet`.  And I think this could be a good dev model for how other experimental features (mixing, DLCs, stuff like that), could be built on `bdk`.

rajarshimaitra commented:
Hello @ulrichard , I haven't looked into the PR in detail yet, but judging from the overall objective is it something similar to [BIP322](https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki)? That BIP is a generic message signing protocol using specially formed Bitcoin transactions. And one of the major applications of such generic signing is "Proof of Reserve", as outlined [here](https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki#full-proof-of-funds). 

My understanding is BIP322 can be used for many things including proof of reserve, so maybe you might wanna try to include that standard in your implementation?

Unfortunately, there is no working implementation of BIP322 and `rust-miniscript` had a feature request open for it for some time now.  As far as my understanding, nobody in Blockstream has time to write it, but they would appreciate PRs. 

I have worked on a very rough sketch of it a few months ago trying to layout how basic BIP322 rust structures might look like. https://github.com/rust-bitcoin/rust-miniscript/pull/240

It's not complete but it does basic BIP322 functions of signing and verifying. That might be something of your interest? 

I am also happy to put in hands with you finishing BIP322 (it has been stale for some time now) and BDK can simply use that (it already has `rust-miniscript` dependency), if that aligns with your need.

Overall I feel this magic should reside in some upstream rust library like `rust-miniscript` or `rust-bitcoin` instead of wallet tools. Wallet tools like BDK can then simply expose simple APIs to use it.

ulrichard commented:
> Hello @ulrichard , I haven't looked into the PR in detail yet, but judging from the overall objective is it something similar to [BIP322](https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki)? That BIP is a generic message signing protocol using specially formed Bitcoin transactions. And one of the major applications of such generic signing is "Proof of Reserve", as outlined [here](https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki#full-proof-of-funds).
> 
> My understanding is BIP322 can be used for many things including proof of reserve, so maybe you might wanna try to include that standard in your implementation?
> 
> Unfortunately, there is no working implementation of BIP322 and `rust-miniscript` had a feature request open for it for some time now. As far as my understanding, nobody in Blockstream has time to write it, but they would appreciate PRs.
> 
> I have worked on a very rough sketch of it a few months ago trying to layout how basic BIP322 rust structures might look like. [rust-bitcoin/rust-miniscript#240](https://github.com/rust-bitcoin/rust-miniscript/pull/240)
> 
> It's not complete but it does basic BIP322 functions of signing and verifying. That might be something of your interest?
> 
> I am also happy to put in hands with you finishing BIP322 (it has been stale for some time now) and BDK can simply use that (it already has `rust-miniscript` dependency), if that aligns with your need.
> 
> Overall I feel this magic should reside in some upstream rust library like `rust-miniscript` or `rust-bitcoin` instead of wallet tools. Wallet tools like BDK can then simply expose simple APIs to use it.

Hi @rajarshimaitra, Yes BIP322 along with BIP127 were the main inspirations. Of which I followed BIP127 more closely. 
I didn't realize BIP322 could be implemented in miniscript. I'll have a look at your PR.

ulrichard committed:
adapted to the updated version of the bitcoin crate

notmandatory commented:
Closing the PR and will work with @ulrichard on his https://github.com/ulrichard/bdk-reserves repo instead.
